File: sdk/dotnet/PrimusSaaS.Identity.Validator/Services/JwksService.cs
using PrimusSaaS.Identity.Validator.Models;
using System.Text.Json;

namespace PrimusSaaS.Identity.Validator.Services;

/// <summary>
/// Service for fetching JWKS from Azure AD.
/// </summary>
public class JwksService
{
    private readonly HttpClient _httpClient;
    private readonly JwksCache? _cache;
    private readonly bool _enableCaching;
    private readonly SemaphoreSlim _fetchLock = new(1, 1);

    /// <summary>
    /// Initializes a new instance of the <see cref="JwksService"/> class.
    /// </summary>
    /// <param name="httpClient">HTTP client for fetching JWKS.</param>
    /// <param name="cache">JWKS cache.</param>
    /// <param name="enableCaching">Set false to bypass caching entirely (useful for tests/local).</param>
    public JwksService(HttpClient? httpClient = null, JwksCache? cache = null, bool enableCaching = true)
    {
        _httpClient = httpClient ?? new HttpClient();
        _enableCaching = enableCaching;
        _cache = enableCaching ? cache ?? new JwksCache() : null;
    }

    /// <summary>
    /// Gets the JWKS from the specified URI, using cache if available.
    /// </summary>
    /// <param name="jwksUri">The JWKS endpoint URI.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The JSON Web Key Set.</returns>
    public async Task<PrimusJsonWebKeySet> GetJwksAsync(
        string jwksUri,
        CancellationToken cancellationToken = default)
    {
        // Try cache first
        var cachedKeySet = _enableCaching ? _cache?.Get(jwksUri) : null;
        if (cachedKeySet != null)
        {
            return cachedKeySet;
        }

        // Fetch from Azure AD
        await _fetchLock.WaitAsync(cancellationToken);
        try
        {
            // Double-check cache after acquiring lock
            cachedKeySet = _enableCaching ? _cache?.Get(jwksUri) : null;
            if (cachedKeySet != null)
            {
                return cachedKeySet;
            }

            var response = await _httpClient.GetAsync(jwksUri, cancellationToken);
            response.EnsureSuccessStatusCode();

            var json = await response.Content.ReadAsStringAsync(cancellationToken);
            var keySet = JsonSerializer.Deserialize<PrimusJsonWebKeySet>(json);

            if (keySet == null)
            {
                throw new InvalidOperationException("Failed to deserialize JWKS.");
            }

            // Cache the key set
            if (_enableCaching && _cache != null)
            {
                _cache.Set(jwksUri, keySet);
            }

            return keySet;
        }
        finally
        {
            _fetchLock.Release();
        }
    }

    /// <summary>
    /// Gets the JWKS for an Azure AD tenant by tenant ID.
    /// </summary>
    /// <param name="tenantId">The Azure AD tenant ID.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>The JSON Web Key Set.</returns>
    public async Task<PrimusJsonWebKeySet> GetJwksForTenantAsync(
        string tenantId,
        CancellationToken cancellationToken = default)
    {
        var jwksUri = $"https://login.microsoftonline.com/{tenantId}/discovery/v2.0/keys";
        return await GetJwksAsync(jwksUri, cancellationToken);
    }

    /// <summary>
    /// Clears the JWKS cache.
    /// </summary>
    public void ClearCache()
    {
        _cache?.Clear();
    }
}

File: sdk/nodejs/primus-identity-validator/src/services/jwksService.ts
import axios, { AxiosInstance } from 'axios';
import { JsonWebKeySet } from '../types';
import { JwksCache } from './jwksCache';

/**
 * Service for fetching and caching JSON Web Key Sets from Azure AD
 */
export class JwksService {
  private readonly httpClient: AxiosInstance;
  private readonly cache: JwksCache;
  private readonly fetchLock: Map<string, Promise<JsonWebKeySet>>;

  constructor(cache?: JwksCache) {
    this.httpClient = axios.create({ timeout: 10000 });
    this.cache = cache ?? new JwksCache();
    this.fetchLock = new Map();
  }

  /**
   * Gets JWKS from the specified URI with caching
   * @param jwksUri JWKS URI from OpenID configuration
   * @returns JSON Web Key Set
   */
  async getJwksAsync(jwksUri: string): Promise<JsonWebKeySet> {
    // Try cache first
    const cached = this.cache.get(jwksUri);
    if (cached) {
      return cached;
    }

    // Check if fetch is in progress
    const inProgress = this.fetchLock.get(jwksUri);
    if (inProgress) {
      return inProgress;
    }

    // Fetch from Azure AD
    const fetchPromise = this.fetchJwks(jwksUri);
    this.fetchLock.set(jwksUri, fetchPromise);

    try {
      const keySet = await fetchPromise;
      
      if (!keySet.keys || keySet.keys.length === 0) {
        throw new Error('No keys found in JWKS');
      }

      // Cache the key set
      this.cache.set(jwksUri, keySet);

      return keySet;
    } finally {
      this.fetchLock.delete(jwksUri);
    }
  }

  /**
   * Gets JWKS for a specific Azure AD tenant
   * @param tenantId Azure AD tenant ID
   * @returns JSON Web Key Set
   */
  async getJwksForTenantAsync(tenantId: string): Promise<JsonWebKeySet> {
    const jwksUri = `https://login.microsoftonline.com/${tenantId}/discovery/v2.0/keys`;
    return this.getJwksAsync(jwksUri);
  }

  /**
   * Fetches JWKS from the specified URI
   * @param uri JWKS URI
   * @returns JSON Web Key Set
   */
  private async fetchJwks(uri: string): Promise<JsonWebKeySet> {
    const response = await this.httpClient.get<JsonWebKeySet>(uri);
    return response.data;
  }

  /**
   * Clears the JWKS cache
   */
  clearCache(): void {
    this.cache.clear();
  }

  /**
   * Gets the number of cached JWKS
   */
  get cacheCount(): number {
    return this.cache.count;
  }
}
